iothub_client/src/iothubtransport_amqp_common.c

```diff
@@ -2492,57 +2419,49 @@ STRING_HANDLE IoTHubTransport_AMQP_Common_GetHostname(TRANSPORT_LL_HANDLE handle
 
-            // Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_09_112: [A DEVICE_MESSAGE_DISPOSITION_INFO instance shall be created with a copy of the `link_name` and `message_id` contained in `message_data`]
-            if ((device_message_disposition_info = create_device_message_disposition_info_from(message_data)) == NULL)
+            /* Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_10_003: [IoTHubTransport_AMQP_Common_SendMessageDisposition shall fail and return IOTHUB_CLIENT_ERROR if the POST message fails, otherwise return IOTHUB_CLIENT_OK.] */
+            if (amqp_device_send_message_disposition(amqp_device_instance->device_handle, (DEVICE_MESSAGE_DISPOSITION_INFO*)device_message_disposition_info, device_disposition_result) != RESULT_OK)
             {
-                // Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_09_113: [If the DEVICE_MESSAGE_DISPOSITION_INFO fails to be created, `IoTHubTransport_AMQP_Common_SendMessageDisposition()` shall fail and return IOTHUB_CLIENT_ERROR]
-                LogError("Device '%s' failed sending message disposition (failed creating DEVICE_MESSAGE_DISPOSITION_RESULT)", STRING_c_str(message_data->transportContext->device_state->device_id));
+                LogError("Device '%s' failed sending message disposition (amqp_device_send_message_disposition failed)", STRING_c_str(amqp_device_instance->device_id));
                 result = IOTHUB_CLIENT_ERROR;
             }
             else
             {
-                /* Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_10_003: [IoTHubTransport_AMQP_Common_SendMessageDisposition shall fail and return IOTHUB_CLIENT_ERROR if the POST message fails, otherwise return IOTHUB_CLIENT_OK.] */
-                if (amqp_device_send_message_disposition(message_data->transportContext->device_state->device_handle, device_message_disposition_info, device_disposition_result) != RESULT_OK)
-                {
-                    LogError("Device '%s' failed sending message disposition (amqp_device_send_message_disposition failed)", STRING_c_str(message_data->transportContext->device_state->device_id));
-                    result = IOTHUB_CLIENT_ERROR;
-                }
-                else
-                {
-                    IoTHubMessage_Destroy(message_data->messageHandle);
-                    result = IOTHUB_CLIENT_OK;
-                }
-
-                // Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_09_114: [`IoTHubTransport_AMQP_Common_SendMessageDisposition()` shall destroy the DEVICE_MESSAGE_DISPOSITION_INFO instance]
-                destroy_device_message_disposition_info(device_message_disposition_info);
+                result = IOTHUB_CLIENT_OK;
             }
         }
 
-        MESSAGE_CALLBACK_INFO_Destroy(message_data);
+        // Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_09_114: [`IoTHubTransport_AMQP_Common_SendMessageDisposition()` shall destroy the DEVICE_MESSAGE_DISPOSITION_INFO instance]
+        IoTHubMessage_Destroy(message_handle);
```

Comment:
If `handle==NULL` but `message_handle != NULL`, we will not execute this block (because of `handle==NULL`) but we need to because we have responsibility for freeing any non-NULL `message_handle`.

Because `IoTHubMessage_Destroy` will happily ignore NULL passed into it, I think we should unconditionally call `IoTHubMessage_Destroy` here in this function at the end so we're guaranteed to free(), especially if parameter checking becomes more complex in future.

Reply:
Indeed. That's all I can say. Thanks.